ludwig/config_validation/checks.py

```diff
@@ -0,0 +1,327 @@
+"""Checks that are not easily covered by marshmallow JSON schema validation like parameter interdependencies.
+
+Assumes incoming configs are comprehensive (all parameters and defaults filled in), and has been schema-validated.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Callable
+
+from ludwig.api_annotations import DeveloperAPI
+from ludwig.constants import (
+    AUDIO,
+    BACKEND,
+    BINARY,
+    CATEGORY,
+    COMBINER,
+    DECODER,
+    ENCODER,
+    IMAGE,
+    IN_MEMORY,
+    INPUT_FEATURES,
+    MODEL_ECD,
+    MODEL_GBM,
+    MODEL_TYPE,
+    NAME,
+    NUMBER,
+    OUTPUT_FEATURES,
+    PREPROCESSING,
+    SEQUENCE,
+    SET,
+    SPLIT,
+    TEXT,
+    TRAINER,
+    TYPE,
+    VECTOR,
+)
+from ludwig.decoders.registry import get_decoder_registry
+from ludwig.encoders.registry import get_encoder_registry
+from ludwig.error import ConfigValidationError
+from ludwig.schema.combiners.utils import get_combiner_registry
+from ludwig.schema.optimizers import optimizer_registry
+from ludwig.types import ModelConfigDict
+from ludwig.utils.metric_utils import get_feature_to_metric_names_map
+
+# Set of all sequence feature types.
+SEQUENCE_OUTPUT_FEATURE_TYPES = {SEQUENCE, TEXT, SET, VECTOR}
+
+
+class ConfigCheckRegistry:
+    """A registry of configuration checks."""
+
+    def __init__(self):
+        self._registry = []
+
+    def register(self, check_fn):
+        self._registry.append(check_fn)
+
+    def check_config(self, config: ModelConfigDict) -> None:
+        for check_fn in self._registry:
+            check_fn(config)
+
+
+_CONFIG_CHECK_REGISTRY = ConfigCheckRegistry()
+
+
+def get_config_check_registry():
+    """Returns the config check registry."""
+    return _CONFIG_CHECK_REGISTRY
+
+
+@DeveloperAPI
+def register_config_check(fn) -> Callable:
+    """Registers a config check function."""
+    _CONFIG_CHECK_REGISTRY.register(fn)
+
+
+class ConfigCheck(ABC):
+    """Checks instances of comprehensive (all parameters and defaults filled in) schema-validated config."""
+
+    @staticmethod
+    @abstractmethod
+    def check(config: ModelConfigDict) -> None:
+        """Checks config for validity."""
+        raise NotImplementedError
+
+
+def check_basic_required_parameters(config: ModelConfigDict) -> None:
```

Comment:
I still wonder if we should consider exposing the auxiliary checks through a separate API so they can be used in the FE as well. Flagging here to not lose track of that thread, but this is great.